### PR TITLE
[LW] Keep additional snapshots around

### PR DESCRIPTION
**Pull request by Jolyon-S**
**Goals (and why)**:
There is a rare case where the aggressive snapshot retention causes starting transactions to repeatedly fail:

1. Transactions started, return lock watch version N -> N + X; update not yet gone through cache
2. Other update received; cache now on version N + Y (where Y < X), and has retentioned all snapshots before N + Y.
3. Update from 1. is processed; this throws if there are no snapshots for every version from N to N + X.
4. Transaction retried

This shouldn't reoccur often, but it seems to be for some services.

**Implementation Description (bullets)**:
Keep a minimum number of snapshots in memory. This will mitigate the chances of the above happening significantly, as the retention won't be so aggressive.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Add new tests around this edge case.

**Concerns (what feedback would you like?)**:
Memory impact should be small (only 1000 additional snapshots at most). I just want to ensure that my use of fancy map types is fine.

**Where should we start reviewing?**:
SnapshotStoreImpl

**Priority (whenever / two weeks / yesterday)**:
Today

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImpl.java [hunk 1]
```diff
@@ -27,23 +29,36 @@
...
 import javax.annotation.concurrent.NotThreadSafe;
 
 @NotThreadSafe
 final class SnapshotStoreImpl implements SnapshotStore {
     private static final SafeLogger log = SafeLoggerFactory.get(SnapshotStoreImpl.class);
-    private static final int MAXIMUM_SIZE = 20_000;
+    private static final Sequence MAX_VERSION = Sequence.of(Long.MAX_VALUE);
 
-    private final Map<Sequence, ValueCacheSnapshot> snapshotMap;
-    private final SetMultimap<Sequence, StartTimestamp> liveSequences;
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImpl.java:40]: I don't know if it mattered that this was a set multimap or not, really (or rather, enough that I needed to keep it typed)

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImplTest.java [hunk 2]
```diff
@@ -97,8 +99,50 @@ public void removeTimestampRemovesSnapshotWhenThereAreNoMoreLiveTimestampsForSeq
...
+        removeSnapshotAndAssert(TIMESTAMP_2, SEQUENCE_2).isEmpty();
+        removeSnapshotAndAssert(TIMESTAMP_3, SEQUENCE_3).hasValue(SNAPSHOT_3);
+        removeSnapshotAndAssert(TIMESTAMP_4, SEQUENCE_4).hasValue(SNAPSHOT_4);
+
+        snapshotStore.storeSnapshot(SEQUENCE_5, ImmutableSet.of(TIMESTAMP_5), SNAPSHOT_5);
+        removeSnapshotAndAssert(TIMESTAMP_5, SEQUENCE_5).hasValue(SNAPSHOT_5);
+        assertThat(snapshotStore.getSnapshotForSequence(SEQUENCE_3)).isEmpty();
+    }
+
+    private OptionalAssert<ValueCacheSnapshot> removeSnapshotAndAssert(StartTimestamp timestamp, Sequence sequence) {
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImplTest.java:125]: fancy!

**Review by gmaretic — APPROVED** [review]
Thanks for fixing this!
